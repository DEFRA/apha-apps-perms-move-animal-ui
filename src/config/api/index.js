import { disinfectantConfig } from './disinfectant.js'

export const apiConfig = {
  disinfectant: disinfectantConfig
}
