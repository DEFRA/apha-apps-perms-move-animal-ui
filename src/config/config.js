import convict from 'convict'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const dirname = path.dirname(fileURLToPath(import.meta.url))

const fourHoursMs = 14400000
const oneWeekMs = 604800000

const isProduction = process.env.NODE_ENV === 'production'
const isTest = process.env.NODE_ENV === 'test'
const isDevelopment = process.env.NODE_ENV === 'development'

export const config = convict({
  serviceVersion: {
    doc: 'The service version, this variable is injected into your docker container in CDP environments',
    format: String,
    nullable: true,
    default: null,
    env: 'SERVICE_VERSION'
  },
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test'],
    default: 'development',
    env: 'NODE_ENV'
  },
  port: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3000,
    env: 'PORT'
  },
  staticCacheTimeout: {
    doc: 'Static cache timeout in milliseconds',
    format: Number,
    default: oneWeekMs,
    env: 'STATIC_CACHE_TIMEOUT'
  },
  serviceName: {
    doc: 'Applications Service Name',
    format: String,
    default: 'Get permission to move animals under disease controls'
  },
  root: {
    doc: 'Project root',
    format: String,
    default: path.resolve(dirname, '../..')
  },
  assetPath: {
    doc: 'Asset path',
    format: String,
    default: '/public',
    env: 'ASSET_PATH'
  },
  isProduction: {
    doc: 'If this application running in the production environment',
    format: Boolean,
    default: isProduction
  },
  isDevelopment: {
    doc: 'If this application running in the development environment',
    format: Boolean,
    default: isDevelopment
  },
  isTest: {
    doc: 'If this application running in the test environment',
    format: Boolean,
    default: isTest
  },
  log: {
    enabled: {
      doc: 'Is logging enabled',
      format: Boolean,
      default: process.env.NODE_ENV !== 'test',
      env: 'LOG_ENABLED'
    },
    level: {
      doc: 'Logging level',
      format: ['fatal', 'error', 'warn', 'info', 'debug', 'trace', 'silent'],
      default: 'info',
      env: 'LOG_LEVEL'
    },
    format: {
      doc: 'Format to output logs in.',
      format: ['ecs', 'pino-pretty'],
      default: isProduction ? 'ecs' : 'pino-pretty',
      env: 'LOG_FORMAT'
    },
    redact: {
      doc: 'Log paths to redact',
      format: Array,
      default: isProduction
        ? ['req.headers.authorization', 'req.headers.cookie', 'res.headers']
        : ['req', 'res', 'responseTime']
    }
  },
  httpProxy: /** @type {SchemaObj<string | null>} */ ({
    doc: 'HTTP Proxy',
    format: String,
    nullable: true,
    default: null,
    env: 'CDP_HTTP_PROXY'
  }),
  httpsProxy: /** @type {SchemaObj<string | null>} */ ({
    doc: 'HTTPS Proxy',
    format: String,
    nullable: true,
    default: null,
    env: 'CDP_HTTPS_PROXY'
  }),
  isSecureContextEnabled: {
    doc: 'Enable Secure Context',
    format: Boolean,
    default: isProduction,
    env: 'ENABLE_SECURE_CONTEXT'
  },
  isMetricsEnabled: {
    doc: 'Enable metrics reporting',
    format: Boolean,
    default: isProduction,
    env: 'ENABLE_METRICS'
  },
  session: {
    cache: {
      engine: {
        doc: 'backend cache is written to',
        format: ['redis', 'memory'],
        default: isProduction ? 'redis' : 'memory',
        env: 'SESSION_CACHE_ENGINE'
      },
      name: {
        doc: 'server side session cache name',
        format: String,
        default: 'session',
        env: 'SESSION_CACHE_NAME'
      },
      ttl: {
        doc: 'server side session cache ttl',
        format: Number,
        default: fourHoursMs,
        env: 'SESSION_CACHE_TTL'
      }
    },
    cookie: {
      ttl: {
        doc: 'Session cookie ttl',
        format: Number,
        default: fourHoursMs,
        env: 'SESSION_COOKIE_TTL'
      },
      password: {
        doc: 'session cookie password',
        format: String,
        default: 'the-password-must-be-at-least-32-characters-long',
        env: 'SESSION_COOKIE_PASSWORD',
        sensitive: true
      },
      secure: {
        doc: 'set secure flag on cookie',
        format: Boolean,
        default: isProduction,
        env: 'SESSION_COOKIE_SECURE'
      }
    }
  },
  fileUpload: {
    uploaderUrl: {
      format: String,
      default: null,
      nullable: true,
      env: 'UPLOADER_URL'
    },
    bucket: {
      format: String,
      default: null,
      nullable: true,
      env: 'FILE_S3_BUCKET'
    },
    path: {
      format: String,
      default: null,
      nullable: true,
      env: 'FILE_S3_PATH'
    }
  },
  redis: /** @type {Schema<RedisConfig>} */ ({
    host: {
      doc: 'Redis cache host',
      format: String,
      default: '127.0.0.1',
      env: 'REDIS_HOST'
    },
    username: {
      doc: 'Redis cache username',
      format: String,
      default: '',
      env: 'REDIS_USERNAME'
    },
    password: {
      doc: 'Redis cache password',
      format: '*',
      default: '',
      sensitive: true,
      env: 'REDIS_PASSWORD'
    },
    keyPrefix: {
      doc: 'Redis cache key prefix name used to isolate the cached results across multiple clients',
      format: String,
      default: 'apha-apps-perms-move-animal-ui:',
      env: 'REDIS_KEY_PREFIX'
    },
    useSingleInstanceCache: {
      doc: 'Connect to a single instance of redis instead of a cluster.',
      format: Boolean,
      default: !isProduction,
      env: 'USE_SINGLE_INSTANCE_CACHE'
    },
    useTLS: {
      doc: 'Connect to redis using TLS',
      format: Boolean,
      default: isProduction,
      env: 'REDIS_TLS'
    }
  }),
  nunjucks: {
    watch: {
      doc: 'Reload templates when they are changed.',
      format: Boolean,
      default: isDevelopment
    },
    noCache: {
      doc: 'Use a cache and recompile templates each time',
      format: Boolean,
      default: isDevelopment
    }
  },
  notify: {
    templateId: /** @type {SchemaObj<string | null>} */ {
      format: String,
      default: null,
      nullable: true,
      env: 'NOTIFY_TEMPLATE_ID'
    },
    apiKey: /** @type {SchemaObj<string | null>} */ {
      format: String,
      default: null,
      nullable: true,
      env: 'NOTIFY_API_KEY'
    },
    caseDeliveryEmailAddress: /** @type {SchemaObj<string | null>} */ {
      format: String,
      default: null,
      nullable: true,
      env: 'NOTIFY_CASE_DELIVERY_EMAIL_ADDRESS'
    },
    timeout: {
      doc: 'Timeout for notify requests in milliseconds',
      format: Number,
      default: 10_000,
      env: 'NOTIFY_TIMEOUT'
    }
  },
  featureFlags: {
    biosecurity: {
      doc: 'Feature flag to enable biosecurity sections',
      format: Boolean,
      default: !isProduction,
      env: 'BIOSECURITY_FEATURE_ENABLED'
    }
  }
})

config.validate({ allowed: 'strict' })

/**
 * @import { Schema, SchemaObj } from 'convict'
 * @import { RedisConfig } from '~/src/server/common/helpers/redis-client.js'
 */
