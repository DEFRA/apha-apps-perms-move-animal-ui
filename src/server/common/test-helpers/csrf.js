import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
