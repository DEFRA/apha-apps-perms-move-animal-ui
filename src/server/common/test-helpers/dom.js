import { JSDOM } from 'jsdom'

/**
 * @param {string} payload
 * @returns {Document}
 */
export const parseDocument = (payload) => new JSDOM(payload).window.document
