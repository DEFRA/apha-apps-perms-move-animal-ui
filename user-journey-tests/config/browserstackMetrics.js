const defaultMaximum = 6

export const getBrowserStackRunners = async () => {
  const username = process.env.BROWSERSTACK_USER
  const accessKey = process.env.BROWSERSTACK_KEY

  try {
    const response = await fetch(
      'https://api.browserstack.com/automate/plan.json',
      {
        headers: {
          Authorization: 'Basic ' + btoa(`${username}:${accessKey}`)
        }
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data = await response.json()
    const {
      parallel_sessions_max_allowed: threadsAllowed,
      parallel_sessions_running: threadsRunning
    } = data
    const availableThreads = threadsAllowed - threadsRunning
    const threadsToUse = Math.floor(availableThreads / 2)

    if (threadsToUse < defaultMaximum && threadsToUse < availableThreads) {
      return threadsToUse
    }

    if (defaultMaximum > availableThreads) {
      return availableThreads
    }

    return defaultMaximum
  } catch (error) {
    return defaultMaximum // default to a speculative value
  }
}
