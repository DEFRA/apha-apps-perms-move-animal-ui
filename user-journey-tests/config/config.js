export default {
  logLevel: 'info',
  maxInstance: 5,
  bail: 0,
  specFileRetries: 0,
  specFileRetriesDeferred: true,
  waitForTimeout: 10000,
  waitforInterval: 200,
  connectionRetryTimeout: 120000,
  connectionRetryCount: 3
}
