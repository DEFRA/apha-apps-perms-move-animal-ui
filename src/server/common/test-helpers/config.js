import { config } from '~/src/config/config.js'

/* global jest */

const configGet = config.get.bind(config)

/**
 * @param {string} key
 * @param {any} value
 */
export const spyOnConfig = (key, value) => {
  jest.spyOn(config, 'get').mockImplementation((name) => {
    if (name === key) {
      return value
    } else {
      return configGet(name)
    }
  })
  return config
}
