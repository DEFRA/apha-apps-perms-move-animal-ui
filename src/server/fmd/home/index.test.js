import { homePage } from './index.js'
import { spyOnConfig } from '../../common/test-helpers/config.js'
import { withCsrfProtection } from '../../common/test-helpers/csrf.js'
import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import SessionTestHelper from '~/src/server/common/test-helpers/session-helper.js'

/** @import { Server } from '@hapi/hapi' */

const pageUrl = '/fmd'

describe('HomePage', () => {
  const page = homePage

  it('should have the right key', () => {
    expect(page.key).toBe('home')
  })

  it('should have the right section key', () => {
    expect(page.sectionKey).toBe('fmd')
  })

  it('should have the right view', () => {
    expect(page.view).toBe('fmd/home/index.njk')
  })

  it('should have the right url', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the right page title', () => {
    expect(page.pageTitle).toBe('Move animals under disease controls')
  })

  it('should have the right heading', () => {
    expect(page.heading).toBe('Move animals under disease controls')
  })

  describe('HomePage content checks', () => {
    /** @type {Server} */
    let server
    let session

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    beforeEach(async () => {
      session = await SessionTestHelper.create(server)
    })

    describe('when auth is enabled', () => {
      beforeEach(() => {
        spyOnConfig('featureFlags', {
          authEnabled: true,
          authRequired: true
        })
      })

      it('should match content', async () => {
        const { payload } = await server.inject(
          withCsrfProtection(
            {
              method: 'GET',
              url: pageUrl
            },
            {
              Cookie: session.sessionID
            }
          )
        )

        const content =
          parseDocument(payload).querySelector('#main-content')?.innerHTML

        expect(content).toMatchSnapshot()
      })
    })

    describe('when auth is disabled', () => {
      beforeEach(() => {
        spyOnConfig('featureFlags', {
          authEnabled: false
        })
      })

      it('should match content', async () => {
        const { payload } = await server.inject(
          withCsrfProtection(
            {
              method: 'GET',
              url: pageUrl
            },
            {
              Cookie: session.sessionID
            }
          )
        )

        const content =
          parseDocument(payload).querySelector('#main-content')?.innerHTML

        expect(content).toMatchSnapshot()
      })
    })
  })
})
