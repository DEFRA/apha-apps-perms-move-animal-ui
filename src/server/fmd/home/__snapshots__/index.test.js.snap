// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage HomePage content checks when auth is disabled should match content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <div class="govuk-!-margin-top-0 govuk-!-margin-bottom-8">
          <h1 class="govuk-heading-l">Move animals under disease controls</h1>
        </div>
        <p class="govuk-body">
          Use this service to apply for a movement licence during a foot and
          mouth disease outbreak.
        </p>
        <h3 class="govuk-heading-m">What you can apply for</h3>
        <p class="govuk-body">
          You can use this service to apply for a licence to:
        </p>
        <ul class="govuk-list govuk-list--bullet">
          <li>
            move live animals to slaughter on or off your farm or premises
          </li>
          <li>move carcasses</li>
          <li>move milk products</li>
        </ul>
        <h3 class="govuk-heading-m">Before you start</h3>
        <p class="govuk-body">
          APHA need these details to process your application:
        </p>
        <ul class="govuk-list govuk-list--bullet">
          <li>
            details about the origin, including who is responsible for the farm
            or premises
          </li>
          <li>ID numbers if you are moving animals</li>
          <li>reason and dates for the movement</li>
        </ul>
        <a href="/fmd/task-list" role="button" draggable="false" class="govuk-button govuk-button--start" data-module="govuk-button" data-testid="start-now-btn">
  Start now
  <svg class="govuk-button__start-icon" xmlns="http://www.w3.org/2000/svg" width="17.5" height="19" viewBox="0 0 33 40" aria-hidden="true" focusable="false">
    <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z"></path>
  </svg>
</a>
        <h3 class="govuk-heading-m">Using and sharing your information</h3>
        <p class="govuk-body">
          How we use your personal data is set out in the
          <a href="https://www.gov.uk/government/publications/animal-and-plant-heath-agency-privacy-notices/exotic-notifiable-and-reportable-diseases-privacy-notice#:~:text=APHA%20may%20share%20data%20with,is%20identified%20at%20the%20laboratory." class="govuk-link" target="_blank">exotic, notifiable and reportable diseases privacy notice (opens in
            a new tab)</a>
          .
        </p>
        <h3 class="govuk-heading-m">Get help</h3>
        <p class="govuk-body">
          If you need help using the service you can call 03000 200 301. Lines
          are open from Monday to Friday between 8.30am and 5pm.
        </p>
      </div>
    </div>
  </div>
        "
`;

exports[`HomePage HomePage content checks when auth is enabled should match content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <div class="govuk-!-margin-top-0 govuk-!-margin-bottom-8">
          <h1 class="govuk-heading-l">Move animals under disease controls</h1>
        </div>
        <p class="govuk-body">
          Use this service to apply for a movement licence during a foot and
          mouth disease outbreak.
        </p>
        <h3 class="govuk-heading-m">What you can apply for</h3>
        <p class="govuk-body">
          You can use this service to apply for a licence to:
        </p>
        <ul class="govuk-list govuk-list--bullet">
          <li>
            move live animals to slaughter on or off your farm or premises
          </li>
          <li>move carcasses</li>
          <li>move milk products</li>
        </ul>
        <h3 class="govuk-heading-m">Before you start</h3>
        <p class="govuk-body">
          APHA need these details to process your application:
        </p>
        <ul class="govuk-list govuk-list--bullet">
          <li>
            details about the origin, including who is responsible for the farm
            or premises
          </li>
          <li>ID numbers if you are moving animals</li>
          <li>reason and dates for the movement</li>
        </ul>
          <h3 class="govuk-heading-m">Signing into the service</h3>
          <p class="govuk-body">
            You need a Defra account to use this service. You can register for
            one using your Government Gateway sign in details.
          </p>
          <p class="govuk-body">
            After you sign in you need to enter your contact information, and
            details about your business if you have one.
          </p>
        <a href="/fmd/task-list" role="button" draggable="false" class="govuk-button govuk-button--start" data-module="govuk-button" data-testid="start-now-btn">
  Start now
  <svg class="govuk-button__start-icon" xmlns="http://www.w3.org/2000/svg" width="17.5" height="19" viewBox="0 0 33 40" aria-hidden="true" focusable="false">
    <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z"></path>
  </svg>
</a>
        <h3 class="govuk-heading-m">Using and sharing your information</h3>
        <p class="govuk-body">
          How we use your personal data is set out in the
          <a href="https://www.gov.uk/government/publications/animal-and-plant-heath-agency-privacy-notices/exotic-notifiable-and-reportable-diseases-privacy-notice#:~:text=APHA%20may%20share%20data%20with,is%20identified%20at%20the%20laboratory." class="govuk-link" target="_blank">exotic, notifiable and reportable diseases privacy notice (opens in
            a new tab)</a>
          .
        </p>
        <h3 class="govuk-heading-m">Get help</h3>
        <p class="govuk-body">
          If you need help using the service you can call 03000 200 301. Lines
          are open from Monday to Friday between 8.30am and 5pm.
        </p>
      </div>
    </div>
  </div>
        "
`;
